webstruct/webannotator.py

```diff
@@ -280,11 +303,113 @@ def to_webannotator(tree, entity_colors=None, url=None):
+                                           position=match.start(),
+                                           length=match.end() - match.start(),
+                                           is_tail=is_tail,
+                                           dfs_number=-1))
+
+    if len(ends) != len(starts):
+        raise ValueError('len(ends) != len(starts)')
+
+    # traverse tree in DFS manner
+    # each text is first child
+    # each tail is last
+    ordered = dict()
+    number = 0
+    for action, element in etree.iterwalk(root, events=('start', 'end')):
+
+        if action == 'end':
+            is_tail = True
+            ordered[(element, is_tail)] = number
+            number = number + 1  # for tail
+
+        if action == 'start':
+            is_tail = False
+            ordered[(element, is_tail)] = number
+            number = number + 1  # for text
+            number = number + 1  # for element
+
+    starts = [s for s in translate_to_dfs(starts, ordered)]
+    ends = [e for e in translate_to_dfs(ends, ordered)]
+
+    starts.sort(key=lambda t: (t.dfs_number, t.position))
+    ends.sort(key=lambda t: (t.dfs_number, t.position))
+
+    dfs_order = (max(n for n in ordered.values()) + 1) * [None]
+    for text_node, dfs_number in ordered.items():
+        dfs_order[dfs_number] = text_node
+
+    tasks = []
+    for _id, (start, end) in enumerate(zip(starts, ends)):
+        start_number = start.dfs_number
+        end_number = end.dfs_number
+
+        if start_number == end_number:
+            tasks.append((start, end, _id))
+            continue
+
+        for text_node in dfs_order[start_number + 1:end_number]:
+            if text_node is None:
+                continue
+
+            element, is_tail = text_node
+            is_string = isinstance(element.tag, six.string_types)
+            is_text = isinstance(element.tag, six.text_type)
+
+            if not is_string and not is_text:
+                continue
+
+            if element.tag in ['script', 'style']:
+                continue
+
+            fictive_start = fabricate_start(element, is_tail, start.tag)
+            fictive_end = fabricate_end(element, is_tail, start.tag)
+            tasks.append((fictive_start, fictive_end, _id))
+
+        fictive_end = fabricate_end(start.element, start.is_tail, start.tag)
+        tasks.append((start, fictive_end, _id))
+
+        fictive_start = fabricate_start(end.element, end.is_tail, end.tag)
+        tasks.append((fictive_start, end, _id))
+
+    def byelement(rec): return (rec[0].element, rec[0].is_tail)
+
+    tasks.sort(key=lambda rec: (ordered[byelement(rec)], rec[0].position))
+    for _, enclosures in itertools.groupby(tasks, byelement):
+        enclosures = [e for e in enclosures]
+        enclose(enclosures, entity_colors)
+
+    _copy_title(root)
+    _add_wacolor_elements(root, entity_colors)
     if url is not None:
-        _set_base(tree, url)
-    return tree
+        _set_base(root, url)
+
+    return root
```

Comment:
Could you extracting separate functions from `to_webannotator`? For example, `_copy_title` and `_add_wacolor_elements` calls are definitely not at the same level of abstraction as two etree.iterwalk loops.